File: packages/kotlin/src/main/kotlin/me/really/crypto/X25519.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import java.security.SecureRandom
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters

/**
 * X25519 key agreement backed by BouncyCastle.
 *
 * The package returns the raw 32-byte Diffie-Hellman output. Higher-level
 * protocols must bind it through their own KDF transcript; this primitive does
 * not apply HKDF implicitly because HPKE, MLS, and ratchets label transcripts
 * differently.
 */
public object ReallyMeX25519 {
    public const val SECRET_KEY_LENGTH: Int = 32
    public const val PUBLIC_KEY_LENGTH: Int = 32
    public const val SHARED_SECRET_LENGTH: Int = 32

    /** Generates a random X25519 keypair: 32-byte public key, 32-byte secret. */
    public fun generateKeyPair(): Pair<ByteArray, ByteArray> {
        val secretKey = ByteArray(SECRET_KEY_LENGTH)
        SecureRandom().nextBytes(secretKey)
        return Pair(derivePublicKey(secretKey), secretKey)
    }

    /** Derives an X25519 keypair from a 32-byte secret. */
    public fun deriveKeyPair(secretKey: ByteArray): Pair<ByteArray, ByteArray> =
        Pair(derivePublicKey(secretKey), secretKey.copyOf())

    /** Derives the 32-byte X25519 public key from a 32-byte secret. */
    public fun derivePublicKey(secretKey: ByteArray): ByteArray {
        if (secretKey.size != SECRET_KEY_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return X25519PrivateKeyParameters(secretKey, 0).generatePublicKey().encoded
    }

    /** Derives the raw 32-byte X25519 shared secret. */
    public fun deriveSharedSecret(publicKey: ByteArray, secretKey: ByteArray): ByteArray {
        if (publicKey.size != PUBLIC_KEY_LENGTH || secretKey.size != SECRET_KEY_LENGTH) {
            throw ReallyMeCryptoException.InvalidInput()
        }

        val sharedSecret = ByteArray(SHARED_SECRET_LENGTH)
        try {
            X25519PrivateKeyParameters(secretKey, 0).generateSecret(
                X25519PublicKeyParameters(publicKey, 0),
                sharedSecret,
                0,
            )
        } catch (_: IllegalArgumentException) {
            throw ReallyMeCryptoException.InvalidInput()
        } catch (_: IllegalStateException) {
            throw ReallyMeCryptoException.InvalidInput()
        }

        if (sharedSecret.all { it == 0.toByte() }) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return sharedSecret
    }
}

File: packages/swift/Sources/ReallyMeCrypto/X25519.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import CryptoKit
import Foundation

/// X25519 key agreement backed by CryptoKit.
///
/// The package returns the raw 32-byte Diffie-Hellman output. Higher-level
/// protocols must bind it through their own KDF transcript; this primitive does
/// not apply HKDF implicitly because HPKE, MLS, and ratchets label transcripts
/// differently.
public enum ReallyMeX25519 {
    public static let secretKeyLength = 32
    public static let publicKeyLength = 32
    public static let sharedSecretLength = 32

    /// Generates a random X25519 keypair: 32-byte public key, 32-byte secret.
    public static func generateKeyPair() throws -> (publicKey: [UInt8], secretKey: [UInt8]) {
        let privateKey = Curve25519.KeyAgreement.PrivateKey()
        return (
            publicKey: Array(privateKey.publicKey.rawRepresentation),
            secretKey: Array(privateKey.rawRepresentation)
        )
    }

    /// Derives an X25519 keypair from a 32-byte secret.
    public static func deriveKeyPair(secretKey: [UInt8]) throws -> (publicKey: [UInt8], secretKey: [UInt8]) {
        (
            publicKey: try derivePublicKey(secretKey: secretKey),
            secretKey: secretKey
        )
    }

    /// Derives the 32-byte X25519 public key from a 32-byte secret.
    public static func derivePublicKey(secretKey: [UInt8]) throws -> [UInt8] {
        guard secretKey.count == secretKeyLength else {
            throw ReallyMeCryptoError.invalidInput
        }
        do {
            let privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: Data(secretKey))
            return Array(privateKey.publicKey.rawRepresentation)
        } catch {
            throw ReallyMeCryptoError.invalidInput
        }
    }

    /// Derives the raw 32-byte X25519 shared secret.
    public static func deriveSharedSecret(publicKey: [UInt8], secretKey: [UInt8]) throws -> [UInt8] {
        guard publicKey.count == publicKeyLength,
              secretKey.count == secretKeyLength
        else {
            throw ReallyMeCryptoError.invalidInput
        }
        do {
            let privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: Data(secretKey))
            let peerPublicKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: Data(publicKey))
            let sharedSecret = try privateKey.sharedSecretFromKeyAgreement(with: peerPublicKey)
            var bytes = sharedSecret.withUnsafeBytes { buffer in
                Array(buffer)
            }
            guard bytes.count == sharedSecretLength,
                  bytes.contains(where: { $0 != 0 })
            else {
                ReallyMeCryptoMemory.bestEffortClear(&bytes)
                throw ReallyMeCryptoError.invalidInput
            }
            return bytes
        } catch let error as ReallyMeCryptoError {
            throw error
        } catch {
            throw ReallyMeCryptoError.invalidInput
        }
    }
}
